packages/babel-plugin-jest-hoist/src/index.js

```diff
@@ -8,11 +8,63 @@
 
 'use strict';
 
+function invariant(condition, message) {
+  if (!condition) {
+    throw new Error('babel-plugin-jest-hoist: ' + message);
+  }
+}
+
 const JEST_GLOBAL = {name: 'jest'};
 
+const idVisitor = {
+  ReferencedIdentifier(path) {
+    this.ids.add(path);
+  },
+};
+
 const FUNCTIONS = {
   mock: {
-    checkArgs: args => args.length === 1 && args[0].isStringLiteral(),
+    checkArgs: args => {
+      if (args.length === 1) {
+        return args[0].isStringLiteral();
+      } else if (args.length === 2) {
+        invariant(
+          args[1].isFunction(),
+          'The second argument of `jest.mock` must be a function.'
+        );
+
+        const ids = new Set();
+        args[1].traverse(idVisitor, {ids});
+
+        const outerScope = args[1].parentPath.scope;
+
+        for (let id of ids) {
+          const name = id.node.name;
+          let found = false;
+          let scope = id.scope;
+
+          while (scope !== outerScope) {
+            if (scope.bindings[name]) {
+              found = true;
+              break;
+            }
+
+            scope = scope.parent;
+          }
+
+          if (!found) {
+            invariant(
+              scope.hasGlobal(name) && (name === 'jest' || name === 'require'),
```

Comment:
Oh.. Can something like HTMLElement be here? If yes, maybe we should just allow all globals?